### PR TITLE
fix: Record handled report info during initial write

### PR DESCRIPTION
**Pull request by kattrali**
Before this change, information which is not part of the standard KS
report was serialized in the onCrash callback, which is shared between
notify() and fatal reports, and used simultaneously in the case that
two reports are generated at once.

This change serializes information which applies to only a single report
into the initial report, reserving the onCrash handler for shared global
state which does not become null once set, like session information, and
handlers which are only run in the event of a crash (by definition,
occurring only once).

Fixes a case where handled state can be reported incorrect for notify()
requests as the shared handled state was cleared/set to null by a
concurrent request.

## Tests

The existing tests cover report info serialization and I broke each component at least once. The case which this changeset fixes its hard to boil down to a test case as its the result of a timing issue.

### Outstanding Questions

- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [x] Final review

- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrash.m [hunk 1]
```diff
@@ -345,32 +345,38 @@ - (void)deleteAllReports {
...
-                 lineOfCode:(NSString *)lineOfCode
-                 stackTrace:(NSArray *)stackTrace
-           terminateProgram:(BOOL)terminateProgram {
-    const char *cName = [name cStringUsingEncoding:NSUTF8StringEncoding];
-    const char *cReason = [reason cStringUsingEncoding:NSUTF8StringEncoding];
-    const char *cLanguage =
-        [language cStringUsingEncoding:NSUTF8StringEncoding];
-    const char *cLineOfCode =
-        [lineOfCode cStringUsingEncoding:NSUTF8StringEncoding];
+- (const char *)encodeAsJSONString:(id)object {
```
Comment (fractalwrench) [inline Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrash.m]: It may be worth moving this out to a different source file, as this method doesn't feel exclusively related to creating crash reports, and it'd be good to have this level of error handling anywhere else we're serialising JSON. It would also make it easier to add unit test coverage

File: Source/BugsnagNotifier.m [hunk 2]
```diff
@@ -560,23 +533,21 @@ - (void)notify:(NSString *)exceptionName
...
-    [self.crashSentry reportUserException:reportName reason:reportMessage];
-    bsg_g_bugsnag_data.userOverridesJSON = NULL;
-    bsg_g_bugsnag_data.handledState = NULL;
+    [self.crashSentry reportUserException:reportName
+                                   reason:reportMessage
+                             handledState:[handledState toJson]
+                                 appState:[self.state toDictionary]
+                        callbackOverrides:report.overrides
+                                 metadata:[report.metaData copy]
+                                   config:[self.configuration.config toDictionary]
```
[the remark this answers is not in the document]
Comment (kattrali) [reply Source/BugsnagNotifier.m]: Previously, these values were mutated in this block. Removing the mutation removed the need for a lock.

File: Source/BugsnagNotifier.m [hunk 3]
```diff
@@ -560,23 +533,21 @@ - (void)notify:(NSString *)exceptionName
...
+                             discardDepth:depth];
 
-    // Restore metaData to pre-crash state.
-    [self.metaDataLock unlock];
```
[the remark this answers is not in the document]
Comment (kattrali) [reply Source/BugsnagNotifier.m]: Discussed "offline", though to summarize there isn't a good automated test for this behavior as it depends on timing, though the fix has been verified manually.

File: Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrash.m (hunk 1 above)
Comment (kattrali) [reply Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrash.m]: The method is used in only one place, and is private to this file. I'll move it to the utility function section to make the intent clearer.
